### PR TITLE
feat: implement list command UX

### PR DESCRIPTION
**Pull request by JeyJeyGao**
According to: https://github.com/notaryproject/notation/blob/main/specs/commandline/list.md
Updated `list` command output

Signed-off-by: Junjie Gao <junjiegao@microsoft.com>

### CODE REVIEW

**Review by patrickzheng200 — APPROVED** [review]
LGTM

**Review by jondonas — APPROVED** [review]
LGTM

File: cmd/notation/list.go [hunk 1]
```diff
@@ -55,9 +59,26 @@ func runList(command *cobra.Command, opts *listOpts) error {
...
+}
+
+func output(digest digest.Digest, sigManifests []notationRegistry.SignatureManifest, reference string) error {
+	ref, err := registry.ParseReference(reference)
+	if err != nil {
+		return err
 	}
 
+	sigCount := len(sigManifests)
+	if sigCount > 0 {
```
Comment (shizhMSFT) [inline cmd/notation/list.go:72]: What's the output if `sigCount == 0`?

/cc @yizha1

Comment (yizha1) [reply cmd/notation/list.go:72]: We don't improve the UX for beta.1. No output means success but no signatures.

Comment (JeyJeyGao) [reply cmd/notation/list.go:72]: Consistent with current implementation.

File: cmd/notation/list.go [hunk 2]
```diff
@@ -55,9 +59,26 @@ func runList(command *cobra.Command, opts *listOpts) error {
 	}
 
 	// write out
-	for _, sigManifest := range sigManifests {
-		fmt.Println(sigManifest.Blob.Digest)
+	return output(manifestDesc.Digest, sigManifests, reference)
+}
+
+func output(digest digest.Digest, sigManifests []notationRegistry.SignatureManifest, reference string) error {
```
Comment (shizhMSFT) [inline cmd/notation/list.go:65]: It is better to choose a good name like `printSignatureManifestList` as `output` is too generic and could be easily misused.

File: cmd/notation/list.go [hunk 3]
```diff
@@ -55,9 +59,26 @@ func runList(command *cobra.Command, opts *listOpts) error {
...
+	ref, err := registry.ParseReference(reference)
+	if err != nil {
+		return err
 	}
 
+	sigCount := len(sigManifests)
+	if sigCount > 0 {
+		// print title
+		fmt.Printf("%s/%s@%s\n", ref.Registry, ref.Repository, digest)
```
Comment (shizhMSFT) [inline cmd/notation/list.go:74]: Try
```go
ref.Digest = digest
fmt.Println(ref)
```

File: cmd/notation/list.go [hunk 4]
```diff
@@ -55,9 +59,26 @@ func runList(command *cobra.Command, opts *listOpts) error {
...
+		fmt.Printf("└── %s\n", notationRegistry.ArtifactTypeNotation)
+
+		for _, sigManifest := range sigManifests[:sigCount-1] {
+			// print each signature digest
+			fmt.Printf("    ├── %s\n", sigManifest.Blob.Digest)
+		}
+		fmt.Printf("    └── %s\n", sigManifests[sigCount-1].Blob.Digest)
```
Comment (shizhMSFT) [inline cmd/notation/list.go:81]: @yizha1 In the recent code refactoring according to the OCI Artifact Referrers API, we are no long be able to get the blob digest efficiently. Instead, should we just print out the digest of the signature manifest? /cc @FeynmanZhou

**Review by shizhMSFT — CHANGES_REQUESTED** [review]

File: cmd/notation/list.go (hunk 4 above)
Comment (yizha1) [reply cmd/notation/list.go:81]: > @yizha1 In the recent code refactoring according to the OCI Artifact Referrers API, we are no long be able to get the blob digest efficiently. Instead, should we just print out the digest of the signature manifest? /cc @FeynmanZhou

I am OK with it. This is more aligned with `oras discover` results. Then the list spec should be updated 
Existing
```shell
<registry>/<repository>@<digest>
└── application/vnd.cncf.notary.v2.signature
    ├──<digest_of_signature>
    └──<digest_of_signature>
```
New
```shell
<registry>/<repository>@<digest>
└── application/vnd.cncf.notary.v2.signature
    ├──<digest_of_signature_manifest>
    └──<digest_of_signature_manifest>
```

File: cmd/notation/list.go (hunk 2 above)
Comment (JeyJeyGao) [reply cmd/notation/list.go]: Renamed to `printSignatureManifestList`

File: cmd/notation/list.go (hunk 3 above)
Comment (JeyJeyGao) [reply cmd/notation/list.go]: Tried, but doesn't work.
ref.Digest() is a method which parses ref.Reference to get digest.

File: cmd/notation/list.go (hunk 4 above)
Comment (JeyJeyGao) [reply cmd/notation/list.go]: Updated to digest_of_signature_manifest

**Review by shizhMSFT — APPROVED** [review]
LGTM with suggestions.

File: cmd/notation/list.go [hunk 5]
```diff
@@ -49,15 +57,56 @@ func runList(command *cobra.Command, opts *listOpts) error {
...
+// notation-go refactor.
+func listSignatureManifestDigests(ctx context.Context, manifestDigest digest.Digest, sigRepo *notationregistry.RepositoryClient) ([]digest.Digest, error) {
+	var sigManifestDigests []digest.Digest
+	if err := sigRepo.Repository.Referrers(ctx, ocispec.Descriptor{
+		Digest: manifestDigest,
+	}, notationRegistry.ArtifactTypeNotation, func(referrers []artifactspec.Descriptor) error {
+		for _, desc := range referrers {
+			if desc.MediaType != artifactspec.MediaTypeArtifactManifest {
+				continue
+			}
```
[the remark this answers is not in the document]
Comment (JeyJeyGao) [reply cmd/notation/list.go]: Removed.

File: cmd/notation/list.go [hunk 6]
```diff
@@ -49,15 +57,57 @@ func runList(command *cobra.Command, opts *listOpts) error {
...
-	sigManifests, err := sigRepo.ListSignatureManifests(command.Context(), manifestDesc.Digest)
+	// print all signature manifest digests
+	return printSignatureManifestDigests(command.Context(), manifestDesc.Digest, sigRepo, reference)
+}
+
+// printSignatureManifestDigests returns the signature manifest digest of
+// the subject manifest.
+//
+// TODO: this is a temporary function and will be replaced after
+// notation-go refactor.
```
Comment (priteshbandi) [inline cmd/notation/list.go:68]: Can you please open an issue to track this?

**Review by priteshbandi — APPROVED** [review]
LGTM with a suggestion

File: cmd/notation/list.go [hunk 7]
```diff
@@ -49,15 +57,57 @@ func runList(command *cobra.Command, opts *listOpts) error {
...
+func printSignatureManifestDigests(ctx context.Context, manifestDigest digest.Digest, sigRepo *notationregistry.RepositoryClient, reference string) error {
+	// prepare title
+	ref, err := registry.ParseReference(reference)
 	if err != nil {
-		return fmt.Errorf("lookup signature failure: %v", err)
+		return err
 	}
+	ref.Reference = manifestDigest.String()
+	var titlePrinted = false
```
Comment (shizhMSFT) [inline cmd/notation/list.go:76]:
```suggestion
	titlePrinted := false
```

**Review by shizhMSFT — APPROVED** [review]
LGTM

File: cmd/notation/list.go (hunk 7 above)
Comment (JeyJeyGao) [reply cmd/notation/list.go]: Updated.

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**